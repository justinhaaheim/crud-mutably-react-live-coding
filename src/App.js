import React, { Component } from 'react'
import EditableRow from './EditableRow'
// import PropTypes from 'prop-types'

/* eslint-disable react/no-multi-comp */

// What state does my app need/have?
// When does it change?
//
const baseUrl = 'http://mutably.herokuapp.com/books/'

class App extends Component {
  constructor(props) {
    super(props)

    this.state = {
      books: [],
      loading: true,
    }

    this.updateElement = this.updateElement.bind(this)
    this.deleteElement = this.deleteElement.bind(this)
  }

  componentDidMount() {
    this.fetchAllElements()
  }

  fetchAllElements() {
    return fetch('http://mutably.herokuapp.com/books')
      .then(response => response.json())
      .then((data) => {
        this.setState({
          books: data.books,
          loading: false,
        })
      })
      .catch((err) => {
        console.log('Error during fetch GET request', err)
        this.setState({
          loading: false,
        })
      })
  }

  updateElement(book) {
    const {
      _id, title, author, image,
    } = book

    return fetch(`${baseUrl}${_id}`, {
      method: 'PUT',
      body: JSON.stringify({
        title: title,
        author: author,
        image: image,
      }),
      headers: { 'Content-Type': 'application/json' },
    }).then(console.log)
  }

  deleteElement() {
    return fetch()
    /// more here!
  }

  render() {
    const { books, loading } = this.state

    return (
      <div className="col-sm-8 col-sm-offset-2">
        <nav className="navbar navbar-default">
          <a className="navbar-brand" href="/">
            Mutably Front-end
          </a>
        </nav>
        <div className="panel panel-default">
          <div className="panel-heading">Books!</div>
          {loading ? <div>Loading...</div> : ''}
          <ul className="list-group">
            {books.map(book => (
              <li className="list-group-item" key={book._id}>
                <EditableRow
                  title={book.title}
                  author={book.author}
                  imageUrl={book.image}
                  bookId={book._id}
                  handleSave={this.updateElement}
                  handleDelete={this.deleteElement}
                />
              </li>
            ))}
          </ul>
        </div>
      </div>
    )
  }
}

export default App
