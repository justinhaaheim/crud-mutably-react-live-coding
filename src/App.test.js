import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import App, { EditableRow } from './App'

global.fetch = jest.fn()

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<EditableRow />, div)
})

test('EditableRow changes output when edit is clicked', () => {
  const component = renderer.create(<EditableRow
    title="Test Title"
    author="Test Author"
    bookId={1}
    imageUrl="http://facebook.github.io/jest/img/jest.svg"
  />)

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  // manually trigger the callback
  tree.props.toggleEditable()
  // re-rendering
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  //
  // // manually trigger the callback
  // tree.toggleEditable()
  // // re-rendering
  // tree = component.toJSON()
  // expect(tree).toMatchSnapshot()
})
