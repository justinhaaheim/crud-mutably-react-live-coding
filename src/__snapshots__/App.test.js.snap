// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableRow changes output when edit is clicked 1`] = `
<div>
  <div>
    <button
      className="btn btn-success"
      onClick={[Function]}
    >
      Edit
    </button>
    <span>
      Test Title
      Â (
      Test Author
      )
    </span>
  </div>
  <img
    src="http://facebook.github.io/jest/img/jest.svg"
    style={
      Object {
        "height": "50px",
        "width": "50px",
      }
    }
  />
</div>
`;
